File: low-level-design/3-creational-design-patterns/2-factory-pattern.java
# Bad Example
interface Logistics {
    void send();
}

class Road implements Logistics {
    @Override
    public void send() {
        System.out.println("Sending by road logic");
    }
}

class Air implements Logistics {
    @Override
    public void send() {
        System.out.println("Sending by air logic");
    }
}

class LogisticsService {
    public void send(String mode) {
        if(mode.equals("road")) {
            return new Road();
        } else if(mode.equals("air")) {
            return new Air();
        }
    }
}

# Good Example
interface Logistics {
    void send();
}

class Road implements Logistics {
    @Override
    public void send() {
        System.out.println("Sending by road logic");
    }
}

class Air implements Logistics {
    @Override
    public void send() {
        System.out.println("Sending by air logic");
    }
}

class LogisticsFactory {
    public static Logistics getLogistics(String mode) {
        if(mode.equals("road")) {
            return new Road();
        } else if(mode.equals("air")) {
            return new Air();
        } else {
            throw new IllegalArgumentException("Invalid mode: " + mode);
        }
    }
}

class LogisticsService {
    public void send(String mode) {
        Logistics logistics = LogisticsFactory.getLogistics(mode);
        logistics.send();
    }
}

File: low-level-design/3-creational-design-patterns/2-factory-pattern.py
from abc import ABC, abstractmethod

# Bad Example
class Logistics(ABC):
    @abstractmethod
    def send(self):
        pass

class Road(Logistics):
    def send(self):
        print("Sending by road logic")

class Air(Logistics):
    def send(self):
        print("Sending by air logic")

class LogisticsService:
    def send(self, mode):
        if mode == "road":
            logistics = Road()
            logistics.send()
        elif mode == "air":
            logistics = Air()
            logistics.send()
        else:
            raise ValueError(f"Invalid mode: {mode}")

# Good Example
class Logistics(ABC):
    @abstractmethod
    def send(self):
        pass

class Road(Logistics):
    def send(self):
        print("Sending by road logic")

class Air(Logistics):
    def send(self):
        print("Sending by air logic")

class LogisticsFactory:
    @staticmethod
    def get_logistics(mode):
        if mode == "road":
            return Road()
        elif mode == "air":
            return Air()
        else:
            raise ValueError(f"Invalid mode: {mode}")

class LogisticsService:
    def send(self, mode):
        logistics = LogisticsFactory.getLogistics(mode)
        logistics.send()